File: minigit-5/ts/src/refs.ts
// Reference management (HEAD, branches, tags)

import * as fs from 'fs';
import * as path from 'path';
import { getMinigitDir, ensureDir } from './utils';
import { expandShortSha, getCommit, readObject, getTreeFromTreeIsh, parseTag } from './objects';

export function getHead(repoRoot?: string): string {
  const minigitDir = getMinigitDir(repoRoot);
  const headPath = path.join(minigitDir, 'HEAD');
  return fs.readFileSync(headPath, 'utf8').trim();
}

export function setHead(value: string, repoRoot?: string): void {
  const minigitDir = getMinigitDir(repoRoot);
  const headPath = path.join(minigitDir, 'HEAD');
  fs.writeFileSync(headPath, value + '\n');
}

export function isDetachedHead(repoRoot?: string): boolean {
  const head = getHead(repoRoot);
  return !head.startsWith('ref:');
}

export function getCurrentBranch(repoRoot?: string): string | null {
  const head = getHead(repoRoot);
  if (head.startsWith('ref:')) {
    const ref = head.slice(5).trim();
    if (ref.startsWith('refs/heads/')) {
      return ref.slice(11);
    }
    return ref;
  }
  return null; // Detached HEAD
}

export function getHeadCommit(repoRoot?: string): string | null {
  const head = getHead(repoRoot);

  if (head.startsWith('ref:')) {
    const ref = head.slice(5).trim();
    return resolveRef(ref, repoRoot);
  }

  // Detached HEAD - head is the SHA
  return head;
}

export function resolveRef(ref: string, repoRoot?: string): string | null {
  const minigitDir = getMinigitDir(repoRoot);
  const refPath = path.join(minigitDir, ref);

  if (fs.existsSync(refPath)) {
    const content = fs.readFileSync(refPath, 'utf8').trim();
    // Could be a symbolic ref or a SHA
    if (content.startsWith('ref:')) {
      return resolveRef(content.slice(5).trim(), repoRoot);
    }
    return content;
  }

  return null;
}

export function updateRef(ref: string, sha: string, repoRoot?: string): void {
  const minigitDir = getMinigitDir(repoRoot);
  const refPath = path.join(minigitDir, ref);
  ensureDir(path.dirname(refPath));
  fs.writeFileSync(refPath, sha + '\n');
}

export function deleteRef(ref: string, repoRoot?: string): boolean {
  const minigitDir = getMinigitDir(repoRoot);
  const refPath = path.join(minigitDir, ref);

  if (fs.existsSync(refPath)) {
    fs.unlinkSync(refPath);
    return true;
  }

  return false;
}

export function refExists(ref: string, repoRoot?: string): boolean {
  const minigitDir = getMinigitDir(repoRoot);
  const refPath = path.join(minigitDir, ref);
  return fs.existsSync(refPath);
}

export function createBranch(name: string, sha: string, repoRoot?: string): void {
  const ref = `refs/heads/${name}`;
  if (refExists(ref, repoRoot)) {
    throw new Error(`A branch named '${name}' already exists.`);
  }
  updateRef(ref, sha, repoRoot);
}

export function deleteBranch(name: string, repoRoot?: string): void {
  const ref = `refs/heads/${name}`;
  if (!deleteRef(ref, repoRoot)) {
    throw new Error(`Branch '${name}' not found.`);
  }
}

export function renameBranch(oldName: string, newName: string, repoRoot?: string): void {
  const oldRef = `refs/heads/${oldName}`;
  const newRef = `refs/heads/${newName}`;

  const sha = resolveRef(oldRef, repoRoot);
  if (!sha) {
    throw new Error(`Branch '${oldName}' not found.`);
  }

  if (refExists(newRef, repoRoot)) {
    throw new Error(`A branch named '${newName}' already exists.`);
  }

  updateRef(newRef, sha, repoRoot);
  deleteRef(oldRef, repoRoot);

  // Update HEAD if we renamed the current branch
  const head = getHead(repoRoot);
  if (head === `ref: ${oldRef}`) {
    setHead(`ref: ${newRef}`, repoRoot);
  }
}

export function listBranches(repoRoot?: string): string[] {
  const minigitDir = getMinigitDir(repoRoot);
  const headsDir = path.join(minigitDir, 'refs', 'heads');

  if (!fs.existsSync(headsDir)) {
    return [];
  }

  return listRefsRecursive(headsDir, '');
}

function listRefsRecursive(baseDir: string, prefix: string): string[] {
  const result: string[] = [];
  const entries = fs.readdirSync(baseDir, { withFileTypes: true });

  for (const entry of entries) {
    const name = prefix ? `${prefix}/${entry.name}` : entry.name;
    if (entry.isDirectory()) {
      result.push(...listRefsRecursive(path.join(baseDir, entry.name), name));
    } else {
      result.push(name);
    }
  }

  return result;
}

export function createTag(name: string, sha: string, repoRoot?: string): void {
  const ref = `refs/tags/${name}`;
  if (refExists(ref, repoRoot)) {
    throw new Error(`Tag '${name}' already exists.`);
  }
  updateRef(ref, sha, repoRoot);
}

export function deleteTag(name: string, repoRoot?: string): void {
  const ref = `refs/tags/${name}`;
  if (!deleteRef(ref, repoRoot)) {
    throw new Error(`Tag '${name}' not found.`);
  }
}

export function listTags(repoRoot?: string): string[] {
  const minigitDir = getMinigitDir(repoRoot);
  const tagsDir = path.join(minigitDir, 'refs', 'tags');

  if (!fs.existsSync(tagsDir)) {
    return [];
  }

  return listRefsRecursive(tagsDir, '');
}

// Resolve a revision to a commit SHA
export function resolveRevision(revision: string, repoRoot?: string): string {
  // Handle special suffixes
  const treeMatch = revision.match(/^(.+)\^\{tree\}$/);
  if (treeMatch) {
    const baseSha = resolveRevision(treeMatch[1], repoRoot);
    return getTreeFromTreeIsh(baseSha, repoRoot);
  }

  // Handle colon path syntax (e.g., HEAD:file.txt)
  const colonMatch = revision.match(/^(.+):(.+)$/);
  if (colonMatch) {
    const baseSha = resolveRevision(colonMatch[1], repoRoot);
    return resolvePathInTree(baseSha, colonMatch[2], repoRoot);
  }

  // Handle parent traversal
  const parentMatch = revision.match(/^(.+)\^(\d*)$/);
  if (parentMatch) {
    const baseSha = resolveRevision(parentMatch[1], repoRoot);
    const parentNum = parentMatch[2] ? parseInt(parentMatch[2], 10) : 1;
    return getParent(baseSha, parentNum, repoRoot);
  }

  // Handle ancestor traversal (HEAD~2)
  const ancestorMatch = revision.match(/^(.+)~(\d+)$/);
  if (ancestorMatch) {
    let sha = resolveRevision(ancestorMatch[1], repoRoot);
    const count = parseInt(ancestorMatch[2], 10);
    for (let i = 0; i < count; i++) {
      sha = getParent(sha, 1, repoRoot);
    }
    return sha;
  }

  // Handle HEAD
  if (revision === 'HEAD') {
    const sha = getHeadCommit(repoRoot);
    if (!sha) {
      throw new Error('HEAD does not point to a commit');
    }
    return sha;
  }

  // Try as branch name
  const branchSha = resolveRef(`refs/heads/${revision}`, repoRoot);
  if (branchSha) {
    return branchSha;
  }

  // Try as tag name
  const tagSha = resolveRef(`refs/tags/${revision}`, repoRoot);
  if (tagSha) {
    // If tag points to a tag object, resolve to the commit
    try {
      const { type, content } = readObject(tagSha, repoRoot);
      if (type === 'tag') {
        const tag = parseTag(content);
        return tag.object;
      }
    } catch {
      // Not a valid object, ignore
    }
    return tagSha;
  }

  // Try as SHA (full or abbreviated)
  try {
    return expandShortSha(revision, repoRoot);
  } catch {
    throw new Error(`Unknown revision: ${revision}`);
  }
}

function getParent(sha: string, parentNum: number, repoRoot?: string): string {
  const commit = getCommit(sha, repoRoot);
  if (parentNum < 1 || parentNum > commit.parents.length) {
    throw new Error(`Commit ${sha.slice(0, 7)} has no parent ${parentNum}`);
  }
  return commit.parents[parentNum - 1];
}

function resolvePathInTree(commitOrTreeSha: string, filePath: string, repoRoot?: string): string {
  const { getTree, readObject } = require('./objects');
  const { type, content } = readObject(commitOrTreeSha, repoRoot);

  let treeSha: string;
  if (type === 'commit') {
    const { parseCommit } = require('./objects');
    const commit = parseCommit(content);
    treeSha = commit.tree;
  } else if (type === 'tree') {
    treeSha = commitOrTreeSha;
  } else {
    throw new Error(`Cannot resolve path in ${type}`);
  }

  const parts = filePath.split('/');
  let currentTreeSha = treeSha;

  for (let i = 0; i < parts.length; i++) {
    const part = parts[i];
    const entries = getTree(currentTreeSha, repoRoot);
    const entry = entries.find((e: any) => e.name === part);

    if (!entry) {
      throw new Error(`Path '${filePath}' does not exist`);
    }

    if (i === parts.length - 1) {
      return entry.sha;
    }

    if (entry.mode !== '40000') {
      throw new Error(`'${parts.slice(0, i + 1).join('/')}' is not a directory`);
    }

    currentTreeSha = entry.sha;
  }

  return currentTreeSha;
}

export function getSymbolicRef(ref: string, repoRoot?: string): string | null {
  const minigitDir = getMinigitDir(repoRoot);
  const refPath = path.join(minigitDir, ref);

  if (!fs.existsSync(refPath)) {
    return null;
  }

  const content = fs.readFileSync(refPath, 'utf8').trim();
  if (content.startsWith('ref:')) {
    return content.slice(5).trim();
  }

  return null;
}

export function setSymbolicRef(ref: string, target: string, repoRoot?: string): void {
  const minigitDir = getMinigitDir(repoRoot);
  const refPath = path.join(minigitDir, ref);
  ensureDir(path.dirname(refPath));
  fs.writeFileSync(refPath, `ref: ${target}\n`);
}

File: minigit-5/py/refs.py
"""Reference management for HEAD, branches, and tags."""

from pathlib import Path
from typing import Optional, List, Tuple
import re

from utils import git_dir, resolve_short_sha, read_object


def read_head(repo_root: Path) -> str:
    """Read HEAD content (either 'ref: refs/...' or a SHA)."""
    head_path = repo_root / '.minigit' / 'HEAD'
    return head_path.read_text().strip()


def write_head(repo_root: Path, content: str):
    """Write to HEAD."""
    head_path = repo_root / '.minigit' / 'HEAD'
    head_path.write_text(content + '\n')


def get_head_ref(repo_root: Path) -> Optional[str]:
    """Get the symbolic ref HEAD points to (e.g., refs/heads/main)."""
    head = read_head(repo_root)
    if head.startswith('ref: '):
        return head[5:]
    return None


def is_head_detached(repo_root: Path) -> bool:
    """Check if HEAD is detached."""
    return get_head_ref(repo_root) is None


def resolve_head(repo_root: Path) -> Optional[str]:
    """Resolve HEAD to a commit SHA."""
    head = read_head(repo_root)
    if head.startswith('ref: '):
        ref_path = head[5:]
        return read_ref(repo_root, ref_path)
    else:
        # Detached HEAD - head is the SHA
        return head


def read_ref(repo_root: Path, ref_path: str) -> Optional[str]:
    """Read a reference file and return the SHA."""
    full_path = repo_root / '.minigit' / ref_path
    if full_path.exists():
        return full_path.read_text().strip()
    return None


def write_ref(repo_root: Path, ref_path: str, sha: str):
    """Write a SHA to a reference file."""
    full_path = repo_root / '.minigit' / ref_path
    full_path.parent.mkdir(parents=True, exist_ok=True)
    full_path.write_text(sha + '\n')


def delete_ref(repo_root: Path, ref_path: str):
    """Delete a reference file."""
    full_path = repo_root / '.minigit' / ref_path
    if full_path.exists():
        full_path.unlink()


def list_branches(repo_root: Path) -> List[str]:
    """List all branch names."""
    heads_dir = repo_root / '.minigit' / 'refs' / 'heads'
    if not heads_dir.exists():
        return []
    return sorted([f.name for f in heads_dir.iterdir() if f.is_file()])


def list_tags(repo_root: Path) -> List[str]:
    """List all tag names."""
    tags_dir = repo_root / '.minigit' / 'refs' / 'tags'
    if not tags_dir.exists():
        return []
    return sorted([f.name for f in tags_dir.iterdir() if f.is_file()])


def get_current_branch(repo_root: Path) -> Optional[str]:
    """Get current branch name or None if detached."""
    ref = get_head_ref(repo_root)
    if ref and ref.startswith('refs/heads/'):
        return ref[11:]
    return None


def resolve_ref(repo_root: Path, ref_name: str) -> Optional[str]:
    """Resolve a ref name to a SHA, trying various paths."""
    # Direct SHA
    if re.match(r'^[0-9a-f]{40}$', ref_name):
        return ref_name

    # Short SHA
    if re.match(r'^[0-9a-f]{4,39}$', ref_name):
        full = resolve_short_sha(ref_name, repo_root)
        if full:
            return full

    # HEAD
    if ref_name == 'HEAD':
        return resolve_head(repo_root)

    # Try refs/heads/<name>
    sha = read_ref(repo_root, f'refs/heads/{ref_name}')
    if sha:
        return sha

    # Try refs/tags/<name>
    sha = read_ref(repo_root, f'refs/tags/{ref_name}')
    if sha:
        # May be annotated tag - resolve to commit
        return resolve_tag_to_commit(repo_root, sha)

    # Try refs/<name>
    sha = read_ref(repo_root, f'refs/{ref_name}')
    if sha:
        return sha

    # Try as direct ref path
    sha = read_ref(repo_root, ref_name)
    if sha:
        return sha

    return None


def resolve_tag_to_commit(repo_root: Path, sha: str) -> str:
    """If SHA is a tag object, resolve to its target commit."""
    try:
        obj_type, _ = read_object(sha, repo_root)
        if obj_type == 'tag':
            from objects import Tag
            tag = Tag.read(sha, repo_root)
            return resolve_tag_to_commit(repo_root, tag.object_sha)
        return sha
    except:
        return sha


def resolve_revision(repo_root: Path, revision: str) -> Optional[str]:
    """Resolve a revision expression to a SHA."""
    # Handle HEAD^{tree}
    if revision.endswith('^{tree}'):
        base = revision[:-7]
        commit_sha = resolve_revision(repo_root, base)
        if commit_sha:
            from objects import Commit
            commit = Commit.read(commit_sha, repo_root)
            return commit.tree_sha
        return None

    # Handle <ref>:<path> for file content
    if ':' in revision and not revision.startswith(':'):
        ref_part, path = revision.split(':', 1)
        commit_sha = resolve_revision(repo_root, ref_part)
        if commit_sha:
            return resolve_path_in_commit(repo_root, commit_sha, path)
        return None

    # Handle parent traversal: HEAD^, HEAD~N
    if '^' in revision or '~' in revision:
        return resolve_parent_syntax(repo_root, revision)

    return resolve_ref(repo_root, revision)


def resolve_parent_syntax(repo_root: Path, revision: str) -> Optional[str]:
    """Resolve parent syntax like HEAD^, HEAD~2."""
    # Split on ^ or ~
    parts = re.split(r'(\^|\~)', revision)

    if not parts:
        return None

    # First part is the base ref
    current = resolve_ref(repo_root, parts[0])
    if not current:
        return None

    i = 1
    while i < len(parts):
        op = parts[i]
        i += 1

        if op == '^':
            # Get parent
            num = 1
            if i < len(parts) and parts[i].isdigit():
                num = int(parts[i])
                i += 1

            from objects import Commit
            commit = Commit.read(current, repo_root)
            if num == 0:
                # ^0 means the commit itself
                pass
            elif len(commit.parents) >= num:
                current = commit.parents[num - 1]
            else:
                return None

        elif op == '~':
            # Ancestor
            num = 1
            if i < len(parts) and parts[i].isdigit():
                num = int(parts[i])
                i += 1

            from objects import Commit
            for _ in range(num):
                commit = Commit.read(current, repo_root)
                if not commit.parents:
                    return None
                current = commit.parents[0]

    return current


def resolve_path_in_commit(repo_root: Path, commit_sha: str, path: str) -> Optional[str]:
    """Resolve a path within a commit to its blob/tree SHA."""
    from objects import Commit, Tree

    commit = Commit.read(commit_sha, repo_root)
    tree_sha = commit.tree_sha

    if not path or path == '':
        return tree_sha

    parts = path.split('/')
    current_sha = tree_sha

    for part in parts:
        if not part:
            continue
        tree = Tree.read(current_sha, repo_root)
        found = False
        for entry in tree.entries:
            if entry.name == part:
                current_sha = entry.sha
                found = True
                break
        if not found:
            return None

    return current_sha


def update_head(repo_root: Path, sha: str):
    """Update HEAD to point to a SHA (either via branch ref or detached)."""
    head_ref = get_head_ref(repo_root)
    if head_ref:
        # Update the branch ref
        write_ref(repo_root, head_ref, sha)
    else:
        # Detached HEAD - update HEAD directly
        write_head(repo_root, sha)
